File: 0110-balanced-binary-tree/0110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        
        ## recursive solution
        
        def dfs(root):
            ## if null condition
            if not root: return [True, 0]
            
            ## now get the left and right nodes of tree
            left, right = dfs(root.left), dfs(root.right)
            
            ## calculate balanced (true or false)
            balanced = (left[0] and right[0] and abs(left[1] - right[1]) <= 1)
            
            ## returns boolean and height
            return [balanced, 1+max(left[1], right[1])]
        
        
    
        ## return boolean         
        return dfs(root)[0]    

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        // Recursive solution
        
        // Helper function for depth-first search
        std::function<std::pair<bool, int>(TreeNode*)> dfs = [&](TreeNode* node) -> std::pair<bool, int> {
            // If the node is null
            if (!node) return {true, 0};
            
            // Get the left and right nodes of the tree
            auto left = dfs(node->left);
            auto right = dfs(node->right);
            
            // Calculate if the tree is balanced (true or false)
            bool balanced = (left.first && right.first && abs(left.second - right.second) <= 1);
            
            // Return a pair of boolean and height
            return {balanced, 1 + std::max(left.second, right.second)};
        };
        
        // Return the boolean result
        return dfs(root).first;
    }
};